allow selection of seat type during manual enrollment on instructor d…

zubair-arbi opened the pull request:
…ashboard

SOL-2107
@hasnain-naveed @ibrahimahmed443 @malikshahzad228 @douglashall 
- Add drop down field for selection of available course modes on page `Membership (Batch Enrollment)`
- Update the method `instructor.views.api.students_update_enrollment` to use the value for selected course mode for enrollment
- Add tests for students enrollment from instructor dashboard
- Add jasmine tests for `Batch Enrollment` 

**Sandbox:** https://zubair-arbi.sandbox.edx.org/courses/course-v1:Test+TEST_01+2016/instructor#view-membership

<img width="1119" alt="screen shot 2016-11-03 at 2 14 27 pm" src="https://cloud.githubusercontent.com/assets/5072991/19961298/b5611fee-a1d3-11e6-9342-77659206aed5.png">

douglashall commented on lms/djangoapps/instructor/views/instructor_dashboard.py:
```diff
@@ -467,6 +467,8 @@ def _section_course_info(course, access):
     """ Provide data for the corresponding dashboard section """
     course_key = course.id
+    course_modes = CourseMode.modes_for_course_dict(course_key)
+
```
Remove extra whitespace.

douglashall commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -572,6 +574,7 @@ such that the value can be defined later than this assignment (file load order).
             this.$identifier_input.val('');
             this.$reason_field.val('');
             this.$checkbox_emailstudents.attr('checked', true);
+            this.$course_mode[0].selectedIndex=0;
```
`this.$course_mode.find('option:eq(0)').prop('selected', true);`

douglashall replied:
This is will also resolve your one quality error.

douglashall commented:
@zubair-arbi Looking good! See my comments for changes we need to make. Are there jasmine and/or bok choy tests we can add to in order to cover the new dropdown?

zubair-arbi commented:
@douglashall I have addressed your feedback and added tests (unittest + jasmine) for `Batch Enrollment`.  I have also added sandbox with current updates. Please review.

zubair-arbi commented:
jenkins run bokchoy

douglashall commented on lms/templates/instructor/instructor_dashboard_2/membership.html:
```diff
@@ -50,6 +50,23 @@
+                <p id="auto-enroll-tip">
+                    ${_("The users will be automatically enrolled in the <em>selected</em> course mode.")}
+                    <br /><br />
+                    ${_("Selecting from course modes has no effect if 'Unenroll' is selected.")}
```
"Selecting a course mode has no effect if 'Unenroll' is selected."

douglashall commented:
@zubair-arbi It looks like we still need to account for the case when an email address is used to enroll someone in the course and there is not already a user associated with that email address. I believe the `CourseEnrollmentAllowed` model is involved in this scenario. We may need to a course_mode field to that model as well.

zubair-arbi commented:
@douglashall I have updated model `CourseEnrollmentAllowed` with new field `mode` + migration and added test for the scenario where instructor can enroll unregistered emails in specific course mode using students update enrollment api. Please review.

zubair-arbi commented:
jenkins run js

zubair-arbi commented:
jenkins run bokchoy

douglashall commented:
FYI @srpearce

srpearce commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -1267,11 +1321,102 @@ def test_enroll_with_email_not_registered_autoenroll(self, protocol):
+            "registration form making sure to use {user_email} in the E-mail field.\n"
+            "Once you have registered and activated your account, "
+            "you will see {display_name} listed on your dashboard.\n\n----\n"
+            "This email was automatically sent from edx.org to {user_email}".format(
```
Dear learner _(is it possible to enter a name here?)_,

You have been invited to enroll in {display_name} at edx.org by a member of the course staff.

To enroll in the course, you need to register and activate an account on {site}. Visit {protocol}://{site}/register and complete the registration form. <bold, if possible>Make sure to enter {user_email} in the Email Address field.</bold> 

After you register your {site} account, you will receive an activation email message at {user_email}. Click the link in this message to activate your account.

After you activate your account, you will automatically go to your dashboard, where you will see {display_name} in your list of courses.

Thanks!
The edX team

zubair-arbi replied:
@srpearce We will need a separate ticket to update the email templates since changing the email content was not part of the story.
FYI: @douglashall

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+                </div>
+            </label>
+            <select id="course-modes-list-selector" name="course_mode" class="course-modes-list-selector">
+                <option value="honor">Honor Certificate</option>
```
Honor Code Certificate

zubair-arbi replied:
changed

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+                <div class="hint auto-enroll-hint">
+                    <span class="hint-caret"></span>
+                    <p id="auto-enroll-tip">
+                        The users will be automatically enrolled in the <em>selected</em> course mode.
```
Why the emphasis for "selected" here?

zubair-arbi replied:
Updated

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+        <legend id="heading-batch-enrollment" class="hd hd-3">Batch Enrollment</legend>
+        <label>
+            Enter email addresses and/or usernames separated by new lines or commas.
+            You will not get notification for emails that bounce, so please double-check spelling.
```
You will not receive notifications for messages that cannot be delivered, so be sure to double-check spelling.

zubair-arbi replied:
This is just a fixture and contains content from previous changes so its out of scope of this story.

zubair-arbi replied:
Since these changes are related to enrollment section so I will make changes.

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+                    <p id="auto-enroll-tip">
+                        The users will be automatically enrolled in the <em>selected</em> course mode.
+                        <br><br>
+                        Selecting from course modes has no effect if 'Unenroll' is selected.
```
If you select **Unenroll** below, you do not need to select a course mode.

zubair-arbi replied:
updated

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+                        If this option is <em>checked</em>, users who have not yet registered for Your Platform Name
+                        Here will be automatically enrolled.
+                        If this option is left <em>unchecked</em>, users who have not yet registered for Your Platform
+                        Name Here will not be enrolled, but will be allowed to enroll once they make an account.
```
- Should "Your Platform Name Here" be something like {platform_name}?

Suggested text revisions (note no `<em>` tags):

If you select this option, learners will be automatically enrolled in the course. If a learner does not yet have an account on {site}, the learner will be automatically enrolled in the course when the learner registers and activates an account.

If you do not select this option, learners will not be automatically enrolled in the course. If a learner does not yet have an account on {site}, the learner can enroll in the course after the learner registers and activates an account.

zubair-arbi replied:
updated

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+                        If this option is left <em>unchecked</em>, users who have not yet registered for Your Platform
+                        Name Here will not be enrolled, but will be allowed to enroll once they make an account.
+                        <br><br>
+                        Checking this box has no effect if 'Unenroll' is selected.
```
If you select **Unenroll** below, you do not need to select this option.

zubair-arbi replied:
updated

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+            <label class="has-hint">
+                <input name="auto-enroll" id="auto-enroll" value="Auto-Enroll" checked="checked"
+                       aria-describedby="heading-batch-enrollment" type="checkbox">
+                <span class="label-text">Auto Enroll</span>
```
Enroll learners automatically

zubair-arbi replied:
This is just a fixture for testing purpose and this content is from previous changes so its out of scope of this story.

zubair-arbi replied:
Also this change will have a greater scope since we will also need to update the existing documentation http://edx.readthedocs.io/projects/open-edx-building-and-running-a-course/en/latest/manage_live_course/course_enrollment.html#options-for-enrolling-learners-in-a-course

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+            <label class="has-hint">
+                <input name="email-students" id="email-students" value="Notify-students-by-email" checked="checked"
+                       aria-describedby="heading-batch-enrollment" type="checkbox">
+                <span class="label-text">Notify users by email</span>
```
Notify learners by email

zubair-arbi replied:
changed

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,71 @@
+                <div class="hint email-students-hint">
+                    <span class="hint-caret"></span>
+                    <p id="email-students-tip">
+                        If this option is <em>checked</em>, users will receive an email notification.
```
If you select this option, learners receive an email message when they are successfully enrolled in your course.

zubair-arbi replied:
changed

srpearce commented on lms/templates/instructor/instructor_dashboard_2/membership.html:
```diff
@@ -50,6 +50,24 @@
+            <div class="hint dropdown-hint auto-enroll-hint">
+                <span class="hint-caret"></span>
+                <p id="auto-enroll-tip">
+                    ${_("The users will be automatically enrolled in the <em>selected</em> course mode.")}
```
Learners will be automatically enrolled in the course mode that you select.

(Note no `<em>` tags.)

srpearce commented on lms/templates/instructor/instructor_dashboard_2/membership.html:
```diff
@@ -50,6 +50,24 @@
+                <p id="auto-enroll-tip">
+                    ${_("The users will be automatically enrolled in the <em>selected</em> course mode.")}
+                    <br /><br />
+                    ${_("Selecting a course mode has no effect if 'Unenroll' is selected.")}
```
If you select **Unenroll** below, you do not need to select a course mode.

zubair-arbi commented:
@srpearce I have addressed your comments. Please have another look.

zubair-arbi commented:
jenkins run python

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,76 @@
+            <label for="course-modes-list-selector" class="has-hint">Select a course mode:
+                <select id="course-modes-list-selector" name="course_mode" class="course-modes-list-selector">
+                <option value="honor">Honor Code Certificate</option>
+                <option value="verified">Verified Certificate</option>
```
Can the user select audit mode as well?

zubair-arbi replied:
Yeah. Any active modes related to the course. Since this is fixture so I just added two modes.

srpearce commented on lms/static/js/fixtures/instructor_dashboard/batchenrollment.html:
```diff
@@ -0,0 +1,76 @@
+            <label class="has-hint">
+                <input name="auto-enroll" id="auto-enroll" value="Auto-Enroll" checked="yes"
+                       aria-describedby="heading-batch-enrollment" type="checkbox">
+                <span class="label-text">Auto Enroll</span>
```
Enroll learners automatically

(If we want to leave it as "Auto enroll", make sure "enroll" isn't capitalized)

zubair-arbi replied:
Changed label to `Enroll learners automatically`

zubair-arbi commented:
@srpearce please review again.

srpearce commented:
👍

srpearce commented:
Oh! One more thing: There shouldn't be an "Audit Certificate" option, since learners who audit the course don't receive a certificate. I believe the options should be:

Audit
Honor code certificate
Verified certificate
Professional education

srpearce commented:
Another question that has come up as I've been working on the course author docs for this: Will Open edX users be required to select a course mode as well, or can they leave this option unselected? Does Open edX have the same default course modes as edx.org?

zubair-arbi commented:
@srpearce I have update the screen shot to reflect the updated course mode titles. Please note that all the new course modes added from [commerce api will have default names set according to list  [`Audit`, `Honor Certificate`, `Verified Certificate`, `Professional Education` and `Credit`]](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/commerce/api/v1/models.py#L40-L55). So to change the modes name we need django admin access or update the commerce api method `get_mode_display_name`.

Instructor will have to select from the provided active course modes list and I verified on my local machine that if there is no existing mode for a course then users are enrolled to the default mode `Audit`. As for the default enrollment mode we have builtin logic in [`enroll` method to select which check for default course mode and enrolls user in it if no course mode is provided](https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/models.py#L1224-L1259) so the behavior would be consistent in open edX too (if this change is applied).

zubair-arbi commented:
jenkins run bokchoy
jenkins run python

zubair-arbi commented:
jenkins run bokchoy

srpearce commented:
@zubair-arbi @mjfrey @marjev I have an open question about this: Will the user be required to select an enrollment mode in the drop-down list, or will there be a "None" option? What is the default?

zubair-arbi commented:
@srpearce According to my experience there will never be an empty drop down even if a course has no course modes (Audit mode will be provided by default). In current implementation there is no default mode pre-selected in drop down (first value is selected always) but I think its a good suggestion to set first value to empty option so that default mode can be applied on backend.
@douglashall

srpearce commented:
Thanks, @zubair-arbi . I'll update the docs accordingly.

srpearce commented:
@douglashall Do you have a rough idea of when this will merge and be live? Since there's no release next week, can I assume it will be live the week of November 21?

douglashall commented:
@srpearce We will be making additional changes to this PR based on feedback from product review. I will keep you updated as to when this will merge.

zubair-arbi commented:
@douglashall should I close this PR?

douglashall commented:
@zubair-arbi No, I think we may still want to revisit this at some point.

smarnach commented:
@zubair-arbi We are interested in this feature.  I [rebased this branch on top of current master](https://github.com/open-craft/edx-platform/commit/92b477c8b210d9f4a98e75c2c49f2b3d4f3a0c74), which resulted in a bunch of conflicts, but none of them seemed to be controversial in their resolution.  I tested the result on my devstack, and it seems to work fine.  Do you have any objections against updating this PR with the rebased branch?

zubair-arbi commented:
@smarnach I don't have any objection just rebase current work onto master and test it.
cc @douglashall

smarnach commented:
@zubair-arbi Thanks!  I've pushed the rebased branch.

douglashall commented:
@smarnach There were some concerns from our product team about opening up the ability for users who have access to the instructor dashboard to enroll learners in an upgraded track.

clemente commented:
@zubair-arbi I rebased on top of master to fix the conflicts (SCSS indentation, code in admin.py which was already merged, and migration order). Because I don't have permission to push to your branch, I pushed it here: https://github.com/open-craft/edx-platform/tree/clemente/zub/sol-2107-instructor-dashboard-add-seat-type-selection
You might want to push this to your branch.

srpearce commented:
@grantgoodmanedX

feanil commented:
Closing due to age and lack of activity.
